Sources/SwiftParser/Declarations.swift

```diff
@@ -710,6 +714,10 @@ extension Parser {
       arena: self.arena
     )
   }
+
+  mutating func atWhereClauseListTerminator() -> Bool {
+    return self.at(.leftBrace)
+  }
```

Comment:
As mentioned in https://github.com/swiftlang/swift-syntax/pull/2689#discussion_r1645084232, should we also disallow trailing commas in `where` clauses?

Reply:
@ahoppen My intention was not completely rule out `where` clauses and `associatedtype` because there are valid cases with clear terminator and to let the parser just fail as today when there’s not.

Reply:
I understand there’s a inconsistency here and I'm fine with going another direction but I was thinking that a user of the language with no knowledge of these ambiguous cases would be confuse with the lack of support in these unambiguous cases so I choose the "do the best we can" approach.

Reply:
It will fail in a where clause for function in a protocol declaration:

```swift
protocol Foo {
  func f<T1, T2>(a: T1, b: T2) where T1: P1, T2: P2, // ❌ Expected type
}
```

But these are more cases with a clear terminator and the parser can handle without ambiguity:

```swift
struct Foo: P1, P2, where P1: Equatable, P2: Equatable, { }

func foo<T1, T2,>() where T1: Equatable, T2: Equatable, { }
      
extension Foo where T1: Equatable, T2: Equatable, { }
```

So there is a inconsistency for `where` clause but we would consistently allow trailing comma for lists with a clear terminator throughout the language.

Reply:
Ah, thanks for the explanation. Makes sense 👍🏽